corehq/sql_db/shard_data_management.py

```diff
@@ -99,6 +99,54 @@ def get_count_of_unmatched_models_by_shard(database, model):
     return results
 
 
+def delete_unmatched_shard_data(database, model, query_size=5000):
+    """
+    Deletes any `model` data residing on shards that aren't expected to live in `database`.
+
+    The deletes are batched in query_size and sorted by the model's partition_attr.
+
+    Returns a generator where each value is a tuple of last value included in the query and number of records deleted
+    """
+    column_name = model.partition_attr
+    shard_id_function = _get_shard_id_function(model)
+
+    # Intentionally uses BETWEEN because it is inclusive of both values
+    query = """
+    DELETE FROM {table_name}
+    WHERE NOT ({shard_id_function}) = ANY(%s) AND {column_name} BETWEEN %s AND %s;
```

Comment:
Since shards have to be consecutive you could use `{shard_id_function} >= %(first_shard)s and {shard_id_function} <= %(last_shard)s`. 

That might be a more efficient query.

Reply:
Do you mean they have to be consecutive on the disk? My main reason for using the partitioned column like this is that it will be efficiently indexed. I had another thought that we could create an index on `shard_id_function` and use that for efficient filtering, but it would add another step to the process